// xdl/build/internal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ANONYMOUS_USERNAME", {
  enumerable: true,
  get: function () {
    return _User().ANONYMOUS_USERNAME;
  }
});
Object.defineProperty(exports, "Analytics", {
  enumerable: true,
  get: function () {
    return _Analytics().default;
  }
});
Object.defineProperty(exports, "AnalyticsClient", {
  enumerable: true,
  get: function () {
    return _Analytics().AnalyticsClient;
  }
});
exports.AndroidCredentials = exports.Android = void 0;
Object.defineProperty(exports, "AnyStandaloneContext", {
  enumerable: true,
  get: function () {
    return _StandaloneContext().AnyStandaloneContext;
  }
});
Object.defineProperty(exports, "ApiV2", {
  enumerable: true,
  get: function () {
    return _ApiV().default;
  }
});
exports.BundleIdentifier = exports.Binaries = exports.AssetBundle = exports.AppleDevice = void 0;
Object.defineProperty(exports, "Config", {
  enumerable: true,
  get: function () {
    return _Config().default;
  }
});
exports.ConnectionStatus = void 0;
Object.defineProperty(exports, "ConnectionType", {
  enumerable: true,
  get: function () {
    return _User().ConnectionType;
  }
});
exports.Env = exports.EmbeddedAssets = exports.Doctor = exports.DevSession = exports.Detach = exports.CoreSimulator = void 0;
Object.defineProperty(exports, "ErrorCode", {
  enumerable: true,
  get: function () {
    return _ErrorCode().ErrorCode;
  }
});
exports.IosWorkspace = exports.IosPlist = exports.IosCodeSigning = exports.ImageUtils = exports.FsCache = exports.Extract = exports.ExponentTools = exports.ExpoUpdatesManifestHandler = exports.ExpSchema = exports.Exp = void 0;
Object.defineProperty(exports, "LoadedHook", {
  enumerable: true,
  get: function () {
    return _runHook().LoadedHook;
  }
});
Object.defineProperty(exports, "LoadingEvent", {
  enumerable: true,
  get: function () {
    return _LoadingEvent().default;
  }
});
exports.LoadingPageHandler = void 0;
Object.defineProperty(exports, "Log", {
  enumerable: true,
  get: function () {
    return _Logger().Log;
  }
});
Object.defineProperty(exports, "LogRecord", {
  enumerable: true,
  get: function () {
    return _PackagerLogsStream().LogRecord;
  }
});
Object.defineProperty(exports, "LogStream", {
  enumerable: true,
  get: function () {
    return _Logger().LogStream;
  }
});
Object.defineProperty(exports, "LogUpdater", {
  enumerable: true,
  get: function () {
    return _PackagerLogsStream().LogUpdater;
  }
});
Object.defineProperty(exports, "Logger", {
  enumerable: true,
  get: function () {
    return _Logger().default;
  }
});
Object.defineProperty(exports, "LoggerDetach", {
  enumerable: true,
  get: function () {
    return _Logger2().default;
  }
});
exports.ModuleVersion = exports.ManifestHandler = void 0;
Object.defineProperty(exports, "NgrokOptions", {
  enumerable: true,
  get: function () {
    return _resolveNgrok().NgrokOptions;
  }
});
exports.PKCS12Utils = void 0;
Object.defineProperty(exports, "PackagerLogsStream", {
  enumerable: true,
  get: function () {
    return _PackagerLogsStream().default;
  }
});
exports.Prompts = exports.ProjectUtils = exports.ProjectSettings = exports.ProjectAssets = exports.Project = void 0;
Object.defineProperty(exports, "PublishOptions", {
  enumerable: true,
  get: function () {
    return _getPublishExpConfigAsync().PublishOptions;
  }
});
Object.defineProperty(exports, "PublishedProjectResult", {
  enumerable: true,
  get: function () {
    return _publishAsync().PublishedProjectResult;
  }
});
Object.defineProperty(exports, "RobotUser", {
  enumerable: true,
  get: function () {
    return _User().RobotUser;
  }
});
Object.defineProperty(exports, "Semaphore", {
  enumerable: true,
  get: function () {
    return _Semaphore().Semaphore;
  }
});
exports.Simulator = exports.SimControlLogs = exports.SimControl = exports.Session = void 0;
Object.defineProperty(exports, "StandaloneBuildFlags", {
  enumerable: true,
  get: function () {
    return _StandaloneBuildFlags().default;
  }
});
Object.defineProperty(exports, "StandaloneContext", {
  enumerable: true,
  get: function () {
    return _StandaloneContext().default;
  }
});
Object.defineProperty(exports, "StartDevServerOptions", {
  enumerable: true,
  get: function () {
    return _startDevServerAsync().StartOptions;
  }
});
Object.defineProperty(exports, "StartOptions", {
  enumerable: true,
  get: function () {
    return _startDevServerAsync().StartOptions;
  }
});
exports.ThirdParty = exports.TableText = void 0;
Object.defineProperty(exports, "UnifiedAnalytics", {
  enumerable: true,
  get: function () {
    return _UnifiedAnalytics().default;
  }
});
exports.UrlUtils = void 0;
Object.defineProperty(exports, "User", {
  enumerable: true,
  get: function () {
    return _User().User;
  }
});
Object.defineProperty(exports, "UserData", {
  enumerable: true,
  get: function () {
    return _UserSettings().UserData;
  }
});
Object.defineProperty(exports, "UserManager", {
  enumerable: true,
  get: function () {
    return _User().default;
  }
});
Object.defineProperty(exports, "UserManagerInstance", {
  enumerable: true,
  get: function () {
    return _User().UserManagerInstance;
  }
});
Object.defineProperty(exports, "UserSettings", {
  enumerable: true,
  get: function () {
    return _UserSettings().default;
  }
});
exports.WebpackEnvironment = exports.Webpack = exports.Watchman = exports.Versions = void 0;
Object.defineProperty(exports, "XDLError", {
  enumerable: true,
  get: function () {
    return _XDLError().default;
  }
});
exports.Xcode = void 0;
Object.defineProperty(exports, "assertValidProjectRoot", {
  enumerable: true,
  get: function () {
    return _errors().assertValidProjectRoot;
  }
});
Object.defineProperty(exports, "broadcastMessage", {
  enumerable: true,
  get: function () {
    return _startAsync().broadcastMessage;
  }
});
Object.defineProperty(exports, "choosePortAsync", {
  enumerable: true,
  get: function () {
    return _choosePortAsync().choosePortAsync;
  }
});
Object.defineProperty(exports, "createBundlesAsync", {
  enumerable: true,
  get: function () {
    return _createBundlesAsync().createBundlesAsync;
  }
});
Object.defineProperty(exports, "delayAsync", {
  enumerable: true,
  get: function () {
    return _delayAsync().delayAsync;
  }
});
Object.defineProperty(exports, "downloadApkAsync", {
  enumerable: true,
  get: function () {
    return _downloadApkAsync().downloadApkAsync;
  }
});
Object.defineProperty(exports, "downloadAppAsync", {
  enumerable: true,
  get: function () {
    return _downloadAppAsync().downloadAppAsync;
  }
});
Object.defineProperty(exports, "getFreePortAsync", {
  enumerable: true,
  get: function () {
    return _getFreePortAsync().getFreePortAsync;
  }
});
Object.defineProperty(exports, "getPublishExpConfigAsync", {
  enumerable: true,
  get: function () {
    return _getPublishExpConfigAsync().getPublishExpConfigAsync;
  }
});
Object.defineProperty(exports, "ip", {
  enumerable: true,
  get: function () {
    return _ip().default;
  }
});
Object.defineProperty(exports, "isDevClientPackageInstalled", {
  enumerable: true,
  get: function () {
    return _isDevClientPackageInstalled().isDevClientPackageInstalled;
  }
});
Object.defineProperty(exports, "learnMore", {
  enumerable: true,
  get: function () {
    return _TerminalLink().learnMore;
  }
});
Object.defineProperty(exports, "pipeOutputToLogger", {
  enumerable: true,
  get: function () {
    return _Logger2().pipeOutputToLogger;
  }
});
Object.defineProperty(exports, "prepareHooks", {
  enumerable: true,
  get: function () {
    return _runHook().prepareHooks;
  }
});
Object.defineProperty(exports, "printBundleSizes", {
  enumerable: true,
  get: function () {
    return _createBundlesAsync().printBundleSizes;
  }
});
Object.defineProperty(exports, "publishAsync", {
  enumerable: true,
  get: function () {
    return _publishAsync().publishAsync;
  }
});
Object.defineProperty(exports, "resolveEntryPoint", {
  enumerable: true,
  get: function () {
    return _resolveEntryPoint().resolveEntryPoint;
  }
});
Object.defineProperty(exports, "resolveNgrokAsync", {
  enumerable: true,
  get: function () {
    return _resolveNgrok().resolveNgrokAsync;
  }
});
Object.defineProperty(exports, "runHook", {
  enumerable: true,
  get: function () {
    return _runHook().runHook;
  }
});
Object.defineProperty(exports, "startAsync", {
  enumerable: true,
  get: function () {
    return _startAsync().startAsync;
  }
});
Object.defineProperty(exports, "startDevServerAsync", {
  enumerable: true,
  get: function () {
    return _startDevServerAsync().startDevServerAsync;
  }
});
Object.defineProperty(exports, "startExpoServerAsync", {
  enumerable: true,
  get: function () {
    return _startLegacyExpoServerAsync().startExpoServerAsync;
  }
});
Object.defineProperty(exports, "startReactNativeServerAsync", {
  enumerable: true,
  get: function () {
    return _startLegacyReactNativeServerAsync().startReactNativeServerAsync;
  }
});
Object.defineProperty(exports, "startTunnelsAsync", {
  enumerable: true,
  get: function () {
    return _ngrok().startTunnelsAsync;
  }
});
Object.defineProperty(exports, "stopAsync", {
  enumerable: true,
  get: function () {
    return _startAsync().stopAsync;
  }
});
Object.defineProperty(exports, "stopExpoServerAsync", {
  enumerable: true,
  get: function () {
    return _startLegacyExpoServerAsync().stopExpoServerAsync;
  }
});
Object.defineProperty(exports, "stopReactNativeServerAsync", {
  enumerable: true,
  get: function () {
    return _startLegacyReactNativeServerAsync().stopReactNativeServerAsync;
  }
});
Object.defineProperty(exports, "stopTunnelsAsync", {
  enumerable: true,
  get: function () {
    return _ngrok().stopTunnelsAsync;
  }
});
Object.defineProperty(exports, "watchBabelConfigForProject", {
  enumerable: true,
  get: function () {
    return _watchBabelConfig().watchBabelConfigForProject;
  }
});
Object.defineProperty(exports, "writeArtifactSafelyAsync", {
  enumerable: true,
  get: function () {
    return _ArtifactUtils().writeArtifactSafelyAsync;
  }
});

function _Semaphore() {
  const data = require("./utils/Semaphore");

  _Semaphore = function () {
    return data;
  };

  return data;
}

function _Env() {
  const data = _interopRequireWildcard(require("./Env"));

  _Env = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Env", {
  enumerable: true,
  get: function () {
    return _Env();
  }
});

function _CoreSimulator() {
  const data = _interopRequireWildcard(require("./apple/CoreSimulator"));

  _CoreSimulator = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "CoreSimulator", {
  enumerable: true,
  get: function () {
    return _CoreSimulator();
  }
});

function _AppleDevice() {
  const data = _interopRequireWildcard(require("./apple/AppleDevice"));

  _AppleDevice = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "AppleDevice", {
  enumerable: true,
  get: function () {
    return _AppleDevice();
  }
});

function _Config() {
  const data = _interopRequireDefault(require("./Config"));

  _Config = function () {
    return data;
  };

  return data;
}

function _Xcode() {
  const data = _interopRequireWildcard(require("./Xcode"));

  _Xcode = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Xcode", {
  enumerable: true,
  get: function () {
    return _Xcode();
  }
});

function _ConnectionStatus() {
  const data = _interopRequireWildcard(require("./ConnectionStatus"));

  _ConnectionStatus = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ConnectionStatus", {
  enumerable: true,
  get: function () {
    return _ConnectionStatus();
  }
});

function _Logger() {
  const data = _interopRequireWildcard(require("./Logger"));

  _Logger = function () {
    return data;
  };

  return data;
}

function _LoadingEvent() {
  const data = _interopRequireDefault(require("./LoadingEvent"));

  _LoadingEvent = function () {
    return data;
  };

  return data;
}

function _TerminalLink() {
  const data = require("./logs/TerminalLink");

  _TerminalLink = function () {
    return data;
  };

  return data;
}

function _Analytics() {
  const data = _interopRequireWildcard(require("./Analytics"));

  _Analytics = function () {
    return data;
  };

  return data;
}

function _UnifiedAnalytics() {
  const data = _interopRequireDefault(require("./UnifiedAnalytics"));

  _UnifiedAnalytics = function () {
    return data;
  };

  return data;
}

function _watchBabelConfig() {
  const data = require("./start/watchBabelConfig");

  _watchBabelConfig = function () {
    return data;
  };

  return data;
}

function _Android() {
  const data = _interopRequireWildcard(require("./Android"));

  _Android = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Android", {
  enumerable: true,
  get: function () {
    return _Android();
  }
});

function _ApiV() {
  const data = _interopRequireDefault(require("./ApiV2"));

  _ApiV = function () {
    return data;
  };

  return data;
}

function _Binaries() {
  const data = _interopRequireWildcard(require("./Binaries"));

  _Binaries = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Binaries", {
  enumerable: true,
  get: function () {
    return _Binaries();
  }
});

function _EmbeddedAssets() {
  const data = _interopRequireWildcard(require("./EmbeddedAssets"));

  _EmbeddedAssets = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "EmbeddedAssets", {
  enumerable: true,
  get: function () {
    return _EmbeddedAssets();
  }
});

function _ErrorCode() {
  const data = require("./ErrorCode");

  _ErrorCode = function () {
    return data;
  };

  return data;
}

function _Exp() {
  const data = _interopRequireWildcard(require("./Exp"));

  _Exp = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Exp", {
  enumerable: true,
  get: function () {
    return _Exp();
  }
});

function _publishAsync() {
  const data = require("./project/publishAsync");

  _publishAsync = function () {
    return data;
  };

  return data;
}

function _createBundlesAsync() {
  const data = require("./project/createBundlesAsync");

  _createBundlesAsync = function () {
    return data;
  };

  return data;
}

function _getPublishExpConfigAsync() {
  const data = require("./project/getPublishExpConfigAsync");

  _getPublishExpConfigAsync = function () {
    return data;
  };

  return data;
}

function _runHook() {
  const data = require("./project/runHook");

  _runHook = function () {
    return data;
  };

  return data;
}

function _ArtifactUtils() {
  const data = require("./tools/ArtifactUtils");

  _ArtifactUtils = function () {
    return data;
  };

  return data;
}

function _ProjectAssets() {
  const data = _interopRequireWildcard(require("./ProjectAssets"));

  _ProjectAssets = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ProjectAssets", {
  enumerable: true,
  get: function () {
    return _ProjectAssets();
  }
});

function _ProjectSettings() {
  const data = _interopRequireWildcard(require("./ProjectSettings"));

  _ProjectSettings = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ProjectSettings", {
  enumerable: true,
  get: function () {
    return _ProjectSettings();
  }
});

function _Prompts() {
  const data = _interopRequireWildcard(require("./Prompts"));

  _Prompts = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Prompts", {
  enumerable: true,
  get: function () {
    return _Prompts();
  }
});

function _SimControl() {
  const data = _interopRequireWildcard(require("./SimControl"));

  _SimControl = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "SimControl", {
  enumerable: true,
  get: function () {
    return _SimControl();
  }
});

function _SimControlLogs() {
  const data = _interopRequireWildcard(require("./SimControlLogs"));

  _SimControlLogs = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "SimControlLogs", {
  enumerable: true,
  get: function () {
    return _SimControlLogs();
  }
});

function _Simulator() {
  const data = _interopRequireWildcard(require("./Simulator"));

  _Simulator = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Simulator", {
  enumerable: true,
  get: function () {
    return _Simulator();
  }
});

function _ThirdParty() {
  const data = _interopRequireWildcard(require("./ThirdParty"));

  _ThirdParty = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ThirdParty", {
  enumerable: true,
  get: function () {
    return _ThirdParty();
  }
});

function _UrlUtils() {
  const data = _interopRequireWildcard(require("./UrlUtils"));

  _UrlUtils = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "UrlUtils", {
  enumerable: true,
  get: function () {
    return _UrlUtils();
  }
});

function _User() {
  const data = _interopRequireWildcard(require("./User"));

  _User = function () {
    return data;
  };

  return data;
}

function _UserSettings() {
  const data = _interopRequireWildcard(require("./UserSettings"));

  _UserSettings = function () {
    return data;
  };

  return data;
}

function _Versions() {
  const data = _interopRequireWildcard(require("./Versions"));

  _Versions = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Versions", {
  enumerable: true,
  get: function () {
    return _Versions();
  }
});

function _Webpack() {
  const data = _interopRequireWildcard(require("./Webpack"));

  _Webpack = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Webpack", {
  enumerable: true,
  get: function () {
    return _Webpack();
  }
});

function _XDLError() {
  const data = _interopRequireDefault(require("./XDLError"));

  _XDLError = function () {
    return data;
  };

  return data;
}

function _AndroidCredentials() {
  const data = _interopRequireWildcard(require("./credentials/AndroidCredentials"));

  _AndroidCredentials = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "AndroidCredentials", {
  enumerable: true,
  get: function () {
    return _AndroidCredentials();
  }
});

function _PKCS12Utils() {
  const data = _interopRequireWildcard(require("./detach/PKCS12Utils"));

  _PKCS12Utils = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "PKCS12Utils", {
  enumerable: true,
  get: function () {
    return _PKCS12Utils();
  }
});

function _PackagerLogsStream() {
  const data = _interopRequireWildcard(require("./logs/PackagerLogsStream"));

  _PackagerLogsStream = function () {
    return data;
  };

  return data;
}

function _Doctor() {
  const data = _interopRequireWildcard(require("./project/Doctor"));

  _Doctor = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Doctor", {
  enumerable: true,
  get: function () {
    return _Doctor();
  }
});

function _ProjectUtils() {
  const data = _interopRequireWildcard(require("./project/ProjectUtils"));

  _ProjectUtils = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ProjectUtils", {
  enumerable: true,
  get: function () {
    return _ProjectUtils();
  }
});

function _ModuleVersion() {
  const data = _interopRequireWildcard(require("./tools/ModuleVersion"));

  _ModuleVersion = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ModuleVersion", {
  enumerable: true,
  get: function () {
    return _ModuleVersion();
  }
});

function _Detach() {
  const data = _interopRequireWildcard(require("./detach/Detach"));

  _Detach = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Detach", {
  enumerable: true,
  get: function () {
    return _Detach();
  }
});

function _IosCodeSigning() {
  const data = _interopRequireWildcard(require("./detach/IosCodeSigning"));

  _IosCodeSigning = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "IosCodeSigning", {
  enumerable: true,
  get: function () {
    return _IosCodeSigning();
  }
});

function _ip() {
  const data = _interopRequireDefault(require("./ip"));

  _ip = function () {
    return data;
  };

  return data;
}

function _ImageUtils() {
  const data = _interopRequireWildcard(require("./tools/ImageUtils"));

  _ImageUtils = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ImageUtils", {
  enumerable: true,
  get: function () {
    return _ImageUtils();
  }
});

function _Extract() {
  const data = _interopRequireWildcard(require("./Extract"));

  _Extract = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Extract", {
  enumerable: true,
  get: function () {
    return _Extract();
  }
});

function _Session() {
  const data = _interopRequireWildcard(require("./Session"));

  _Session = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Session", {
  enumerable: true,
  get: function () {
    return _Session();
  }
});

function _StandaloneContext() {
  const data = _interopRequireWildcard(require("./detach/StandaloneContext"));

  _StandaloneContext = function () {
    return data;
  };

  return data;
}

function _ExponentTools() {
  const data = _interopRequireWildcard(require("./detach/ExponentTools"));

  _ExponentTools = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ExponentTools", {
  enumerable: true,
  get: function () {
    return _ExponentTools();
  }
});

function _IosPlist() {
  const data = _interopRequireWildcard(require("./detach/IosPlist"));

  _IosPlist = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "IosPlist", {
  enumerable: true,
  get: function () {
    return _IosPlist();
  }
});

function _IosWorkspace() {
  const data = _interopRequireWildcard(require("./detach/IosWorkspace"));

  _IosWorkspace = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "IosWorkspace", {
  enumerable: true,
  get: function () {
    return _IosWorkspace();
  }
});

function _errors() {
  const data = require("./project/errors");

  _errors = function () {
    return data;
  };

  return data;
}

function _ngrok() {
  const data = require("./start/ngrok");

  _ngrok = function () {
    return data;
  };

  return data;
}

function _startDevServerAsync() {
  const data = require("./start/startDevServerAsync");

  _startDevServerAsync = function () {
    return data;
  };

  return data;
}

function _startLegacyExpoServerAsync() {
  const data = require("./start/startLegacyExpoServerAsync");

  _startLegacyExpoServerAsync = function () {
    return data;
  };

  return data;
}

function _startLegacyReactNativeServerAsync() {
  const data = require("./start/startLegacyReactNativeServerAsync");

  _startLegacyReactNativeServerAsync = function () {
    return data;
  };

  return data;
}

function _ExpSchema() {
  const data = _interopRequireWildcard(require("./project/ExpSchema"));

  _ExpSchema = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ExpSchema", {
  enumerable: true,
  get: function () {
    return _ExpSchema();
  }
});

function _delayAsync() {
  const data = require("./utils/delayAsync");

  _delayAsync = function () {
    return data;
  };

  return data;
}

function _choosePortAsync() {
  const data = require("./utils/choosePortAsync");

  _choosePortAsync = function () {
    return data;
  };

  return data;
}

function _downloadApkAsync() {
  const data = require("./utils/downloadApkAsync");

  _downloadApkAsync = function () {
    return data;
  };

  return data;
}

function _downloadAppAsync() {
  const data = require("./utils/downloadAppAsync");

  _downloadAppAsync = function () {
    return data;
  };

  return data;
}

function _isDevClientPackageInstalled() {
  const data = require("./utils/isDevClientPackageInstalled");

  _isDevClientPackageInstalled = function () {
    return data;
  };

  return data;
}

function _BundleIdentifier() {
  const data = _interopRequireWildcard(require("./BundleIdentifier"));

  _BundleIdentifier = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "BundleIdentifier", {
  enumerable: true,
  get: function () {
    return _BundleIdentifier();
  }
});

function _FsCache() {
  const data = _interopRequireWildcard(require("./tools/FsCache"));

  _FsCache = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "FsCache", {
  enumerable: true,
  get: function () {
    return _FsCache();
  }
});

function _WebpackEnvironment() {
  const data = _interopRequireWildcard(require("./webpack-utils/WebpackEnvironment"));

  _WebpackEnvironment = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "WebpackEnvironment", {
  enumerable: true,
  get: function () {
    return _WebpackEnvironment();
  }
});

function _Logger2() {
  const data = _interopRequireWildcard(require("./detach/Logger"));

  _Logger2 = function () {
    return data;
  };

  return data;
}

function _StandaloneBuildFlags() {
  const data = _interopRequireDefault(require("./detach/StandaloneBuildFlags"));

  _StandaloneBuildFlags = function () {
    return data;
  };

  return data;
}

function _AssetBundle() {
  const data = _interopRequireWildcard(require("./detach/AssetBundle"));

  _AssetBundle = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "AssetBundle", {
  enumerable: true,
  get: function () {
    return _AssetBundle();
  }
});

function _TableText() {
  const data = _interopRequireWildcard(require("./logs/TableText"));

  _TableText = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "TableText", {
  enumerable: true,
  get: function () {
    return _TableText();
  }
});

function _resolveEntryPoint() {
  const data = require("./tools/resolveEntryPoint");

  _resolveEntryPoint = function () {
    return data;
  };

  return data;
}

function _Watchman() {
  const data = _interopRequireWildcard(require("./Watchman"));

  _Watchman = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Watchman", {
  enumerable: true,
  get: function () {
    return _Watchman();
  }
});

function _DevSession() {
  const data = _interopRequireWildcard(require("./DevSession"));

  _DevSession = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "DevSession", {
  enumerable: true,
  get: function () {
    return _DevSession();
  }
});

function _resolveNgrok() {
  const data = require("./start/resolveNgrok");

  _resolveNgrok = function () {
    return data;
  };

  return data;
}

function _startAsync() {
  const data = require("./start/startAsync");

  _startAsync = function () {
    return data;
  };

  return data;
}

function _ManifestHandler() {
  const data = _interopRequireWildcard(require("./start/ManifestHandler"));

  _ManifestHandler = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ManifestHandler", {
  enumerable: true,
  get: function () {
    return _ManifestHandler();
  }
});

function _ExpoUpdatesManifestHandler() {
  const data = _interopRequireWildcard(require("./start/ExpoUpdatesManifestHandler"));

  _ExpoUpdatesManifestHandler = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ExpoUpdatesManifestHandler", {
  enumerable: true,
  get: function () {
    return _ExpoUpdatesManifestHandler();
  }
});

function _LoadingPageHandler() {
  const data = _interopRequireWildcard(require("./start/LoadingPageHandler"));

  _LoadingPageHandler = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "LoadingPageHandler", {
  enumerable: true,
  get: function () {
    return _LoadingPageHandler();
  }
});

function _getFreePortAsync() {
  const data = require("./start/getFreePortAsync");

  _getFreePortAsync = function () {
    return data;
  };

  return data;
}

function _Project() {
  const data = _interopRequireWildcard(require("./Project"));

  _Project = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Project", {
  enumerable: true,
  get: function () {
    return _Project();
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
